File: day5/day5.java
import java.util.ArrayList;
import java.util.Scanner;

public class day5 {
    private static int jump(Integer[] numbers, Replacer replacer) {
        int current = 0;
        int last = 0;
        int count = 0;
        while (current >= 0 && current < numbers.length) {
            last = current;
            current += numbers[current];
            numbers[last] = replacer.replace(numbers[last]);
            count += 1;
        }
        return count;
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        ArrayList<Integer> linesArray = new ArrayList<>();
        while (scan.hasNext()) {
            linesArray.add(scan.nextInt());
        }
        Integer[] partOne = linesArray.toArray(new Integer[]{});
        Integer[] partTwo = partOne.clone();

        Replacer addOne = (int oldValue) -> oldValue + 1;
        Replacer stranger = (int oldValue) -> oldValue < 3 ? oldValue + 1 : oldValue -1;

        System.out.println(jump(partOne, addOne));
        System.out.println(jump(partTwo, stranger));
    }
}

interface Replacer {
    int replace(int oldValue);
}

File: day5/day5.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "../api.h"

int jump(int* numbers, int numberCount, int (*replace)(int)) {
    int current, last, count = 0;
    while (current >= 0 && current < numberCount) {
        last = current;
        current += numbers[current];
        numbers[last] = replace(numbers[last]);
        count += 1;
    }
    return count;
}

int add_one(int oldValue) {
	return oldValue + 1;
}

int stranger(int oldValue) {
	if (oldValue >= 3) {
		return oldValue - 1;
	}
	return oldValue + 1;
}

int main(int argc, char** argv) {
	Input* input = load_input();

	int* partOne = malloc(sizeof(int) * input->count);
	for (int i = 0; i < input->count; i++) {
		partOne[i] = atoi(input->lines[i]);
	}

	printf("%d\n", jump(partOne, input->count, add_one));

	int* partTwo = malloc(sizeof(int) * input->count);
	for (int i = 0; i < input->count; i++) {
		partTwo[i] = atoi(input->lines[i]);
	}

	printf("%d\n", jump(partTwo, input->count, stranger));
}